{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5562f629",
   "metadata": {},
   "outputs": [],
   "source": [
    "### In this notebook we add new-primers to the ABE primer list and re-order the library"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "c5cdd378",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd \n",
    "from os import listdir\n",
    "from os.path import join\n",
    "import sys\n",
    "sys.path.append('../src')\n",
    "from utils import *\n",
    "import matplotlib.pyplot as plt\n",
    "import pickle\n",
    "from Bio import SeqIO"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "cb2d5044",
   "metadata": {},
   "outputs": [],
   "source": [
    "out_path = '../../data/output/sg_Finder'\n",
    "mrna_path = '../../data/MANE'\n",
    "fpath = '../../data/output/sg_out/'\n",
    "\n",
    "# Load all expressed transcripts\n",
    "ess_genelist = pd.read_csv('../../data/output/k562_comm_ess.txt', header =None)\n",
    "neutral_genelist = pd.read_csv('../../data/output/k562_neutral.txt', header =None)\n",
    "\n",
    "fpath = '../../data/output/sg_out/'\n",
    "list_gene = [i for i in listdir(fpath) if not i.startswith('.')]\n",
    "list_gene = [i for i in list_gene if len(listdir(join(fpath,i))) > 5]\n",
    "list_gene_ess = [i for i in list_gene if i in list(ess_genelist[0])]\n",
    "list_gene_neutral = [i for i in list_gene if i in list(neutral_genelist[0])]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "c1410e15",
   "metadata": {},
   "outputs": [],
   "source": [
    "# How many genes not have any sgRNA detected from CHOPCHOP\n",
    "# for gene_folder in join(out_path,gene,)\n",
    "# 6 genes are first filtered out\n",
    "inter_gene =list(set(ess_genelist[0]).intersection(set(list_gene_ess)))\n",
    "left_out_genes = [ i for i in list(set(ess_genelist[0]))  if i not in list_gene_ess]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "6d3fb107",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "6 genes didnt have any sgRNAgenerated by CHOPCHOP:['FARSB', 'LRR1', 'RPS9', 'GINS1', 'POLD3', 'SARS1']\n"
     ]
    }
   ],
   "source": [
    "print(f'{len(left_out_genes)} genes didnt have any sgRNA\\\n",
    "generated by CHOPCHOP:{left_out_genes}')\n",
    "### Usually because the isoforms varies too much that CHOPCHOP can't find a union. Tested SARS1 on website version."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "9198659e",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "### Filtered genes. They are the same across 11nt window and 13nt window\n",
    "df_filter = pd.read_csv(join(out_path, 'ess_15','filter_gene.csv'), index_col =0)\n",
    "df_filter = df_filter.set_index('gene')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "dceee2e1",
   "metadata": {},
   "source": [
    "***Get synonymous and stop codon sgRNA for ess and neutral genes***"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "6566c5c0",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Read no primer list\n",
    "ABE_sg = SeqIO.parse('../../data/output/primers/rREsg_ABE.fasta','fasta')\n",
    "abe_sg = [str(i.seq) for i in ABE_sg]\n",
    "# Dual_sg = SeqIO.parse('../data/output/primers/rREsg_Dual.fasta','fasta')\n",
    "# dual_sg = [str(i.seq) for i in Dual_sg]\n",
    "abe_mis_ety = pickle.load(open('ABE_MIS_ETY.pkl','rb'))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "c443fda4",
   "metadata": {},
   "outputs": [],
   "source": [
    "final_abe = abe_sg+abe_mis_ety"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "168c2a8d",
   "metadata": {},
   "source": [
    "## ***Add primers***"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "ffa80d21",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "1739 ABE sg groups \n"
     ]
    }
   ],
   "source": [
    "rABE_noprimer_list,rABE_pass_list = get_noprimer_list(final_abe)\n",
    "print(f'{len(rABE_noprimer_list)} ABE sg groups ')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "abf21274",
   "metadata": {},
   "outputs": [],
   "source": [
    "write_noprime_list('ABE_order2.fasta',final_abe)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "fc6f32af",
   "metadata": {},
   "outputs": [],
   "source": [
    "# F = 'GGGATCACTTACTACTTCCG'\n",
    "# R = 'GGTACCCTATCACTTAACCG'\n",
    "### We use the CBE primer to order this library\n",
    "list_primers = []\n",
    "\n",
    "F = 'CACTAGAGTAATCGCTACCG'\n",
    "R = 'GGGCTACTAGACATAACTCG'\n",
    "\n",
    "list_primers.append(F)\n",
    "list_primers.append(R)\n",
    "\n",
    "Fwd,Rev,oligoTM, primerTM = calc_best_primerTM(F,R, rABE_noprimer_list)\n",
    "ABE_final_seq = attach_group_primer(rABE_pass_list,Fwd[:18],Rev[:18])\n",
    "df_abe_final = pd.DataFrame({'sequence':ABE_final_seq})\n",
    "df_abe_final['sequence'] = df_abe_final['sequence'].apply(lambda x:''.join(x))\n",
    "df_abe_final = df_abe_final.reset_index()\n",
    "df_abe_final['name'] = 'ABE_'+df_abe_final['index'].astype(str)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "97aa2bde",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "ThermoResult(structure_found=True, tm=-11.30, dg=-3947.59, dh=-39400.00, ds=-114.31)\n"
     ]
    }
   ],
   "source": [
    "from itertools import combinations\n",
    "for (i,j) in combinations(list_primers,2):\n",
    "    print(p3.calcHeterodimer(i,j))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "9482641a",
   "metadata": {},
   "outputs": [],
   "source": [
    "df_abe_final[['name','sequence']].to_csv('../data/output/ABE_order.csv',index = False)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.18"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
